utils/sysfs/sysfs.go

```diff
@@ -335,25 +352,205 @@ func (fs *realSysFs) GetSystemUUID() (string, error) {
 	}
 }
 
-func (fs *realSysFs) IsCPUOnline(dir string) bool {
-	cpuPath := fmt.Sprintf("%s/online", dir)
-	content, err := ioutil.ReadFile(cpuPath)
+func (fs *realSysFs) IsCPUOnline(cpuPath string) bool {
+	return isCPUOnline(cpuPath)
+}
+
+// isCPUOnline verifies if a CPU in online. Takes path to /sys/devices/system/cpu/cpu*/ directory as a parameter.
+func isCPUOnline(cpuPath string) bool {
+	if isx86() && isZeroCPU(cpuPath) {
+		// It *is* possible to enable CPU hotplug for CPU0 but it requires setting a boot flag.
+		// It is assumed it's a rare situation.
+		return true
+	}
+
+	const isOnline = 49
+	onlinePath := filepath.Join(cpuPath, "online")
+	onlineVal, err := ioutil.ReadFile(onlinePath)
+	// First, make sure that no error occurred.
 	if err != nil {
-		pathErr, ok := err.(*os.PathError)
-		if ok {
-			if errors.Is(pathErr.Unwrap(), os.ErrNotExist) && isZeroCPU(dir) {
+		// If error occurred and is os.PathError then it might be os.ErrNotExists.
+		pathErr, isPathError := err.(*os.PathError)
+		if isPathError {
+			if !os.IsNotExist(pathErr.Unwrap()) {
+				// If online file exists but reading failed we assume that CPU is offline.
+				klog.V(2).Infof("Cannot determine CPU %s online state, assuming offline", cpuPath)
+				return false
+			}
+			// If online file does not exists then we should check one more location.
+			klog.V(2).Infof("%s not found, checking %s directory", onlinePath, filepath.Join(cpuPath, "hotplug"))
+			if isOnlineHotplug(cpuPath) {
 				return true
 			}
+			klog.V(2).Infof("unable to determine hotplug state of %s CPU; assuming that CPU if offline", cpuPath)
+			return false
 		}
-		klog.Warningf("unable to read %s: %s", cpuPath, err.Error())
+	}
+	// Check if online file value is 1.
+	onlineVal = bytes.TrimSpace(onlineVal)
+	if len(onlineVal) == 0 || onlineVal[0] != isOnline {
+		klog.V(2).Infof("CPU %s is offline", cpuPath)
+		return false
+	}
+	return true
+}
+
+func isx86() bool {
```

Comment:
i'm still a bit confused why this check is needed (see above comment), but if needed, wouldn't it better to use a build tag and have a amd64 and !not amd64 variant of the file or something like that?

Reply:
Using build tag will make testing it a bit difficult unless we can test non-x86_64 build.

Reply:
I was, obviously, wrong. Fixed :)